contracts/libraries/Assessment/AssessmentVoteLib.sol

```diff
@@ -0,0 +1,176 @@
+        "Cannot withdraw rewards from votes which are in lockup period"
+      );
+      IAssessment.Poll memory poll =
+        IAssessment.EventType(vote.eventType) == IAssessment.EventType.CLAIM
+        ? claims[vote.eventId].poll
+        : incidents[vote.eventId].poll;
+
+      totalReward = _getTotalRewardForEvent(
+        CONFIG,
+        IAssessment.EventType(vote.eventType),
+        vote.eventId,
+        claims,
+        incidents
+      );
+      rewardToWithdraw += totalReward * vote.tokenWeight / (poll.accepted + poll.denied);
+    }
+
+    stakeOf[user].rewardsWithdrawnUntilIndex = withdrawUntilIndex;
+    nxm.mint(user, rewardToWithdraw);
+  }
+
+  function castVote (
+    IAssessment.Configuration calldata CONFIG,
+    uint8 eventType,
+    uint104 id,
+    bool accepted,
+    mapping(address => IAssessment.Stake) storage stakeOf,
+    mapping(address => IAssessment.Vote[]) storage votesOf,
+    mapping(bytes32 => bool) storage hasAlreadyVotedOn,
+    IAssessment.Claim[] storage claims,
+    IAssessment.Incident[] storage incidents
+  ) external {
+
+    {
+      bytes32 voteHash = keccak256(abi.encodePacked(id, msg.sender, eventType));
+      require(!hasAlreadyVotedOn[voteHash], "Already voted");
+      hasAlreadyVotedOn[voteHash] = true;
+    }
+
+    IAssessment.Stake memory stake = stakeOf[msg.sender];
+    require(stake.amount > 0, "A stake is required to cast votes");
+
+    uint payoutImpact;
+    IAssessment.Poll memory poll;
+    uint32 blockTimestamp = uint32(block.timestamp);
+    if (IAssessment.EventType(eventType) == IAssessment.EventType.CLAIM) {
+      IAssessment.Claim memory claim = claims[id];
+      poll = claims[id].poll;
+      payoutImpact = AssessmentUtilsLib._getPayoutImpactOfClaim(claim.details);
+      require(blockTimestamp < poll.end, "Voting is closed");
+    } else {
+      IAssessment.Incident memory incident = incidents[id];
+      poll = incidents[id].poll;
+      payoutImpact = AssessmentUtilsLib._getPayoutImpactOfIncident(incident.details);
+      require(blockTimestamp < poll.end, "Voting is closed");
+    }
+
+    require(
+      poll.accepted > 0 || accepted == true,
+      "At least one accept vote is required to vote deny"
+    );
+
+    if (accepted) {
+      if (poll.accepted == 0) {
+        poll.start = blockTimestamp;
+      }
+      poll.accepted += stake.amount;
+    } else {
+      poll.denied += stake.amount;
+    }
+
+    poll.end = AssessmentUtilsLib._calculatePollEndDate(CONFIG, poll, payoutImpact);
+
+    if (poll.end < blockTimestamp) {
+      // When poll end date falls in the past, replace it with the current block timestamp
+      poll.end = blockTimestamp;
+    }
+
+    // [todo] Add condition when vote shifts poll end in the past and write end with the
+    // current blcok timestamp. Could also consider logic where the consensus is shifted at the
+    // very end of the voting period.
+
+    if (IAssessment.EventType(eventType) == IAssessment.EventType.CLAIM) {
```

Comment:
i'm thinking whether incident and claim voting should be separate functions.

while this may save some lines, it makes it a bit harder to reason about each in isolation, also if their logic diverges at some point it can get more complex

Reply:
The purpose of this rewrite is to use the same assessment mechanism for multiple type of events, namely claims and incidents at the moment, allowing the possibility to add more when needed.